<launch>
  <arg name="run_sim" default="1" />

  <!-- Buffer Server -->
  <node pkg="tf2_ros" type="buffer_server" name="tf2_buffer_server_plugs" output="screen">
    <remap from="tf2_buffer_server" to="tf2_buffer_server_plugs" />
    <param name="buffer_size" value="120.0"/>
  </node>

  <rosparam command="load" file="$(find pr2_plugs_actions)/config/joint_trajectories.yaml" />
  <rosparam command="load" file="$(find pr2_plugs_actions)/config/cart_ik_seeds.yaml" />

  <!-- wall norm action -->
  <include file="$(find stereo_wall_detection)/wall_extractor.launch"/>
  <node pkg="pr2_plugs_actions" type="detect_wall_norm.py" name="detect_wall_norm" output="screen"/>

  <!-- vision detect outlet action -->
  <node pkg="pr2_plugs_actions" type="vision_detect_outlet" name="vision_outlet_detection" output="screen">
    <param name="outlet_template" type="str" value="$(find pr2_plugs_actions)/outlet_templates/configs/green_2x1white"/>
  </node>

  <!-- vision detect plug action -->
  <node pkg="pr2_plugs_actions" type="vision_detect_plug" name="vision_plug_detection" output="screen">
    <param name="subpixel_corners" type="bool" value="true"/>		
    <rosparam file="$(find pr2_plugs_actions)/config/plug_description.yaml" />
  </node>

  <!-- detect plug action -->
  <node pkg="pr2_plugs_actions" type="detect_plug.py" name="detect_plug_in_gripper" output="screen"/>

  <!-- detect plug on base action -->
  <node pkg="pr2_plugs_actions" type="detect_plug_on_base.py" name="detect_plug_on_base" output="screen"/>

  <!-- align base action -->
  <node pkg="pr2_plugs_actions" type="align_base" name="align_base" output="screen"/>

  <!-- rechare toplevel action -->
  <node pkg="pr2_plugs_actions" type="recharge_toplevel.py" name="recharge_toplevel" output="screen" />

  <!-- detect outlet action -->
  <node pkg="pr2_plugs_actions" type="detect_outlet.py" name="detect_outlet" output="screen" />
  <rosparam file="$(find pr2_plugs_actions)/config/sim_calibration.yaml"  if="$(arg run_sim)" />
  <rosparam file="/etc/ros/plugs/hw_calibration.yaml" unless="$(arg run_sim)" />

  <!-- fetch plug action -->
  <node pkg="pr2_plugs_actions" type="fetch_plug.py" name="fetch_plug" output="screen" />

  <!-- plug in action -->
  <node pkg="pr2_plugs_actions" type="plug_in.py" name="plug_in" output="screen" />
  
  <!-- stow plug action -->
  <node pkg="pr2_plugs_actions" type="stow_plug.py" name="stow_plug" output="screen"/>

  <!-- image snapshotter action action -->
  <node pkg="pr2_image_snapshot_recorder" type="image_snapshotter.py" name="image_snapshotter" output="screen"/>

  <!-- wiggle plug action -->
  <node pkg="pr2_plugs_actions" type="wiggle_plug.py" name="wiggle_plug" output="screen"/>

  <!-- outlet location server -->
  <rosparam file="$(find pr2_plugs_actions)/config/outlet_approach_poses.yaml"/>	
  <node pkg="pr2_plugs_actions" type="outlet_server.py" name="outlet_server" output="screen"/>

  <!-- move base omnidirectional action -->
  <node pkg="pr2_plugs_actions" type="move_base_omnidirectional" name="move_base_omnidirectional" output="screen">\
    <param name="k_trans" value="5.0" />
    <param name="k_rot" value="5.0" />
    <param name="tolerance_trans" value="0.02" />
    <param name="tolerance_rot" value="0.03" />
  </node>
  <group ns="move_base_omnidirectional/costmap_move_base" >
    <rosparam file="$(find pr2_plugs_actions)/launch/footprint.yaml" command="load" />
    <!-- Start Costmap Parameters -->
    <param name="global_frame" value="odom_combined" />
    <param name="robot_base_frame" value="base_link" />
    <param name="update_frequency" value="5.0" />
    <param name="publish_frequency" value="1.0" />
    <param name="obstacle_range" value="1.0" />
    <param name="max_obstacle_height" value="2.0" />
    <param name="raytrace_range" value="1.5" />
    <param name="observation_sources" value="base_scan" />
    <param name="base_scan/observation_persistence" value="0.0" />
    <param name="base_scan/expected_update_rate" value="0.2" />
    <param name="base_scan/data_type" value="LaserScan" />
    <param name="base_scan/clearing" value="true" />
    <param name="base_scan/marking" value="false" />
    <param name="base_scan_marking/observation_persistence" value="0.0" />
    <param name="base_scan_marking/expected_update_rate" value="0.2" />
    <param name="base_scan_marking/data_type" value="PointCloud" />
    <param name="base_scan_marking/clearing" value="false" />
    <param name="base_scan_marking/marking" value="true" />
    <param name="rolling_window" value="true" />
    <param name="static_map" value="false" />
    <!-- begin only used if static map is set to false -->
    <param name="width" value="5.0" />
    <param name="height" value="5.0" />
    <param name="origin_x" value="-2.5" />
    <param name="origin_y" value="-2.5" />
    <param name="resolution" value="0.01" />
    <!-- end only used if static map is set to false -->
    <param name="cost_scaling_factor" value="1.0" />
    <param name="lethal_cost_threshold" value="100" />
    <!-- End Costmap Parameters -->
    <remap from="obstacles" to="~obstacles" />
    <remap from="inflated_obstacles" to="~inflated_obstacles" />
  </group>

</launch>


